### PR TITLE
Res 682: make Links sparse in HTM regions

### PR DESCRIPTION
**Pull request by breznak**
from upstream https://github.com/numenta/nupic.core/pull/1429

### CODE REVIEW

File: src/nupic/engine/Link.cpp [hunk 1]
```diff
@@ -344,12 +344,22 @@ void Link::compute() {
   }
 
   if (src_->isSparse() == dest_->isSparse()) {
-    // No conversion required, just copy the buffer over
-    ::memcpy((char *)(dest.getBuffer()) + destByteOffset, src.getBuffer(),
```
Comment (breznak) [inline src/nupic/engine/Link.cpp]: always nice to remove this!

File: src/nupic/engine/Link.cpp [hunk 2]
```diff
@@ -344,12 +344,22 @@ void Link::compute() {
...
+      for (size_t i = 0; i < srcCount; i++) {
+        destBuf[destIdx++] = destOffset_ + srcBuf[i];
+      }
+
       // Remove 'const' to update the variable length array
-      const_cast<Array &>(dest).setCount(src.getCount());
+      const_cast<Array &>(dest).setCount(destIdx);
+    } else {
+      // No conversion required, just copy the buffer over
+      ::memcpy((char *)(dest.getBuffer()) + destByteOffset, src.getBuffer(),
```
Comment (breznak) [inline src/nupic/engine/Link.cpp]: nah :D

**Comment by breznak** [comment]
David, this is related to your proposal in #72 , is it a pre-requisite, or the Datatype solution will work completely different way?

**Comment by dkeeney** [comment]
My thinking is that the isSparse thing should be entirely hidden inside the
Array object.
But I am sure this will work.

**Comment by breznak** [comment]
Ok, so we'll leave it to rot a bit here, see if/when numenta merges that to upstream. When we get to this task, we can merge and then do your suggestions above it. I also like the "Sparse/Dense data type" approach better.

**Comment by breznak** [comment]
I'm closing this PR as it's still (and likely won't) be accepted by numenta's upstream.

**Pull request closed without merging**